File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/java/handlers/UIHandler.java
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.java.handlers;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

import static com.zebra.aisuite_quickstart.utils.CommonUtils.WAREHOUSE_LOCALIZER;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureException;
import androidx.camera.core.ImageProxy;

import com.zebra.aisuite_quickstart.filtertracker.FilterDialog;
import com.zebra.aisuite_quickstart.filtertracker.FilterItem;
import com.zebra.aisuite_quickstart.java.CameraXLivePreviewActivity;
import com.zebra.aisuite_quickstart.java.camera.CameraManager;
import com.zebra.aisuite_quickstart.utils.CommonUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * UIHandler manages UI-related operations including spinner setup,
 * overlay management, and orientation changes.
 */
public class UIHandler {
    private static final String TAG = "UIHandler";

    private final CameraXLivePreviewActivity activity;
    private final CameraManager cameraManager;

    // Model constants
    public static final String BARCODE_DETECTION = "Barcode";
    private static final String LEGACY_BARCODE_DETECTION = "Legacy Barcode";
    public static final String TEXT_OCR_DETECTION = "OCR";
    private static final String LEGACY_OCR_DETECTION = "Legacy OCR";
    public static final String ENTITY_ANALYZER = "Tracker";
    public static final String PRODUCT_RECOGNITION = "Product Recognition";
    private static final String LEGACY_PRODUCT_RECOGNITION = "Legacy Product Recognition";
    private static final String ENTITY_VIEW_FINDER = "Entity Viewfinder";

    private String selectedModel = BARCODE_DETECTION;
    private final SharedPreferences sharedPreferences;
    private boolean isEntityViewFinder = false;
    public boolean isSpinnerInitialized = false;
    private ImageCapture imageCapture;
    public boolean isInCaptureMode = false;
    private Bitmap currentCapture;
    private final ExecutorService executors = Executors.newFixedThreadPool(2);

    public UIHandler(CameraXLivePreviewActivity activity, CameraManager cameraManager, SharedPreferences sharedPreferences) {
        this.activity = activity;
        this.sharedPreferences = sharedPreferences;
        this.cameraManager = cameraManager;
    }

    public void setupSpinner() {
        ArrayAdapter<String> dataAdapter = getStringArrayAdapter();
        activity.getBinding().spinner.setAdapter(dataAdapter);
        activity.getBinding().spinner.setOnItemSelectedListener(createSpinnerListener());
        setupFilterButton();
    }

    @NonNull
    private ArrayAdapter<String> getStringArrayAdapter() {
        List<String> options = new ArrayList<>();
        options.add(BARCODE_DETECTION);
        options.add(TEXT_OCR_DETECTION);
        options.add(PRODUCT_RECOGNITION);
        options.add(ENTITY_ANALYZER);
        options.add(ENTITY_VIEW_FINDER);
        options.add(WAREHOUSE_LOCALIZER);
        // options.add(LEGACY_BARCODE_DETECTION); // uncomment to use barcode legacy option
        // options.add(LEGACY_OCR_DETECTION); // uncomment to use ocr legacy option
       // options.add(LEGACY_PRODUCT_RECOGNITION); // Uncomment to use Product recognition legacy option

        ArrayAdapter<String> dataAdapter = new ArrayAdapter<>(activity, com.zebra.aisuite_quickstart.R.layout.spinner_style, options);
        dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        return dataAdapter;
    }

    private AdapterView.OnItemSelectedListener createSpinnerListener() {
        return new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int pos, long l) {
                selectedModel = adapterView.getItemAtPosition(pos).toString();
                isEntityViewFinder = selectedModel.equals(ENTITY_VIEW_FINDER);

                Log.e(TAG, "selected option is " + selectedModel);
                activity.getBinding().trackerFilter.setVisibility(TextUtils.equals(selectedModel, ENTITY_ANALYZER) ? VISIBLE : GONE);

                // Lock orientation when Entity Viewfinder is selected
                if (isEntityViewFinder) {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LOCKED);
                    Log.d(TAG, "Orientation locked for Entity Viewfinder mode");
                } else {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
                    Log.d(TAG, "Orientation unlocked for " + selectedModel + " mode");
                }
                activity.setModelLoaded(false);
                if (selectedModel.equalsIgnoreCase(BARCODE_DETECTION) || selectedModel.equalsIgnoreCase(TEXT_OCR_DETECTION) || selectedModel.equalsIgnoreCase(PRODUCT_RECOGNITION) || selectedModel.equalsIgnoreCase(ENTITY_ANALYZER) || selectedModel.equalsIgnoreCase(WAREHOUSE_LOCALIZER)) {
                    initializeCaptureFeature();
                    activity.getBinding().captureLayout.setVisibility(VISIBLE);
                }  else {
                    activity.getBinding().captureLayout.setVisibility(GONE);
                }

                if (!isSpinnerInitialized) {
                    isSpinnerInitialized = true;
                } else {
                    activity.clearGraphicOverlay();
                    activity.stopAnalyzing();
                    cameraManager.unbindAll();
                    activity.disposeModels();
                    cameraManager.bindPreviewAndAnalysis(activity.getPreviewSurfaceProvider());
                    activity.bindAnalysisUseCase();
                }
                // Clear overlays and rebind camera

            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
                // No action needed
            }
        };
    }

    private void setupFilterButton() {
        activity.getBinding().trackerFilter.setOnClickListener(v -> {
            FilterDialog dialog = new FilterDialog(activity);
            dialog.setCallback(options -> {
                SharedPreferences.Editor editor = sharedPreferences.edit();

                for (FilterItem option : options) {
                    editor.putBoolean(option.getTitle(), option.isChecked());
                }
                editor.apply();
                activity.setModelLoaded(false);

                activity.clearGraphicOverlay();
                activity.stopAnalyzing();
                cameraManager.unbindAll();
                activity.disposeModels();
                cameraManager.bindPreviewAndAnalysis(activity.getPreviewSurfaceProvider());
                activity.bindAnalysisUseCase();
            });
            dialog.show();
        });
    }

    public void updatePreviewVisibility(boolean isEntityViewFinder) {
        activity.getBinding().previewView.setVisibility(isEntityViewFinder ? GONE : VISIBLE);
        activity.getBinding().entityView.setVisibility(isEntityViewFinder ? VISIBLE : GONE);
    }

    public String getSelectedModel() {
        return selectedModel;
    }

    public boolean isEntityViewFinder() {
        return isEntityViewFinder;
    }
    public void setSpinnerEnabled(boolean enabled) {
        if (activity != null && activity.getBinding() != null) {
            activity.runOnUiThread(() -> {
                activity.getBinding().spinner.setEnabled(enabled);
                activity.getBinding().spinner.setAlpha(enabled ? 1.0f : 0.5f);

                activity.getBinding().trackerFilter.setEnabled(enabled);
                activity.getBinding().trackerFilter.setAlpha(enabled ? 1.0f : 0.5f);

                activity.getBinding().captureButton.setEnabled(enabled);
                activity.getBinding().captureButton.setAlpha(enabled ? 1.0f : 0.5f);
            });
        }
    }
    private void initializeCaptureFeature() {
        activity.getBinding().captureButton.setOnClickListener(v -> {
            if(selectedModel.equalsIgnoreCase(BARCODE_DETECTION) || selectedModel.equalsIgnoreCase(TEXT_OCR_DETECTION) || selectedModel.equalsIgnoreCase(PRODUCT_RECOGNITION) || selectedModel.equalsIgnoreCase(ENTITY_ANALYZER) || selectedModel.equalsIgnoreCase(WAREHOUSE_LOCALIZER)) {
                if(activity.isModelLoaded()) {
                    activity.getBinding().captureLayout.setVisibility(GONE);
                    imageCapture = cameraManager.getImageCapture();
                    activity.stopAnalyzing();
                    activity.getBinding().control.setVisibility(GONE);

                    // Submit capture initialization to background thread
                    executors.submit(() -> {
                        if (selectedModel.equalsIgnoreCase(BARCODE_DETECTION)) {
                            initializeCaptureDecoderAndCapture();
                        } else if (selectedModel.equalsIgnoreCase(TEXT_OCR_DETECTION)) {
                            initializeCaptureOCRAndCapture();
                        } else if (selectedModel.equalsIgnoreCase(PRODUCT_RECOGNITION)) {
                            initializeCaptureRecognitionAndCapture();
                        } else if (selectedModel.equalsIgnoreCase(ENTITY_ANALYZER)) {
                            initializeCaptureTracker();
                        } else if (selectedModel.equalsIgnoreCase(WAREHOUSE_LOCALIZER)) {
                            initializeCaptureWareHouseLocalizer();
                        }
                    });
                }else{
                    Toast.makeText(activity, "Model is not loaded", Toast.LENGTH_SHORT).show();
                }
            }
        });
        activity.getBinding().backToLiveButton.setOnClickListener(v -> {
            try {
                activity.getBinding().control.setVisibility(VISIBLE);
                returnToLivePreview();
            } catch (ExecutionException | InterruptedException e) {
                throw new RuntimeException(e);
            }
        });
    }

    private void initializeCaptureWareHouseLocalizer() {
        activity.initializeCaptureWareHouse(this::performCapture);
    }

    private void initializeCaptureOCRAndCapture() {
        activity.initializeCaptureOCR(this::performCapture);
    }

    private void initializeCaptureDecoderAndCapture() {
        activity.initializeCaptureDecoder(this::performCapture);
    }

    private void initializeCaptureRecognitionAndCapture() {
        activity.initializeCaptureRecognition(this::performCapture);
    }

    private void initializeCaptureTracker() {
        activity.initializeCaptureTracker(this::performCapture);
    }

    private void performCapture() {
        if (imageCapture == null || isInCaptureMode) return;
        Log.d(TAG, "Performing capture");
        imageCapture.takePicture(executors, new ImageCapture.OnImageCapturedCallback() {
            @Override
            public void onCaptureSuccess(@NonNull ImageProxy imageProxy) {
                try {
                    Log.e(TAG, "Image DIms w x h="+ imageProxy.getWidth() + " x " + imageProxy.getHeight());
                    Log.e(TAG, "Selected model" + selectedModel);
                    currentCapture = CommonUtils.rotateBitmapIfNeeded(imageProxy);
                    activity.runOnUiThread(() -> {
                        switchToCaptureMode();
                        activity.getBinding().capturedImageView.setImageBitmap(currentCapture);
                    });
                    if(selectedModel.equalsIgnoreCase(TEXT_OCR_DETECTION)) {
                        activity.processCaptureOCR(imageProxy);
                    }else if(selectedModel.equalsIgnoreCase(BARCODE_DETECTION)){
                        activity.processBarcodeWithCaptureDecoder(imageProxy);
                    }else if(selectedModel.equalsIgnoreCase(PRODUCT_RECOGNITION)){
                        activity.processCaptureRecognition(imageProxy);
                    } else if(selectedModel.equalsIgnoreCase(ENTITY_ANALYZER)){
                        activity.processCaptureTracker(imageProxy);
                    }else if(selectedModel.equalsIgnoreCase(WAREHOUSE_LOCALIZER)){
                        activity.processCaptureWareHouseLocalizer(imageProxy);
                    }

                    Log.d(TAG, "bitmap captured");
                } catch (Exception e) {
                    Log.e(TAG,"Exception occurred while capturing "+e.getMessage());
                }
            }

            @Override
            public void onError(@NonNull ImageCaptureException exception) {
                Log.e(TAG, "Capture operation failed", exception);
            }
        });
    }

    @SuppressLint("ResourceAsColor")
    private void returnToLivePreview() throws ExecutionException, InterruptedException {
        activity.runOnUiThread(() -> {
            isInCaptureMode = false;
            currentCapture = null;

            // Clear graphic overlay to remove any previous detection results
            activity.clearGraphicOverlay();

            activity.getBinding().capturedImageView.setVisibility(View.GONE);
            activity.getBinding().capturedImageView.setImageBitmap(null);
            activity.getBinding().backToLiveButton.setVisibility(View.GONE);

            activity.getBinding().trackerFilter.setVisibility(TextUtils.equals(selectedModel, ENTITY_ANALYZER) ? VISIBLE : GONE);
            activity.getBinding().previewView.setVisibility(View.VISIBLE);
            activity.getBinding().previewView.setBackgroundColor(android.R.color.black);
            activity.getBinding().graphicOverlay.setVisibility(View.VISIBLE);
            activity.getBinding().captureLayout.setVisibility(View.VISIBLE);

        });
        cameraManager.unbindAll();
        cameraManager.bindPreviewAndAnalysis(activity.getPreviewSurfaceProvider());
        if(selectedModel.equalsIgnoreCase(TEXT_OCR_DETECTION)) {
            activity.setOCRAnalysis();
        }else if(selectedModel.equalsIgnoreCase(BARCODE_DETECTION)){
            activity.setBarcodeAnalysis();
        } else if(selectedModel.equalsIgnoreCase(PRODUCT_RECOGNITION)){
            activity.clearCaptureListener();
            activity.setRecognitionAnalysis();
        } else if(selectedModel.equalsIgnoreCase(ENTITY_ANALYZER)){
            activity.setTrackerAnalysis();
        } else if(selectedModel.equalsIgnoreCase(WAREHOUSE_LOCALIZER)){
            activity.setWareHouseAnalysis();
        }
    }

    private void switchToCaptureMode() {
        activity.runOnUiThread(() -> {
            isInCaptureMode = true;

            cameraManager.unbindImageAnalysis();

            activity.getBinding().previewView.setVisibility(View.GONE);
            activity.getBinding().graphicOverlay.setVisibility(View.GONE);
            activity.getBinding().captureLayout.setVisibility(View.GONE);

            activity.getBinding().capturedImageView.setVisibility(View.VISIBLE);
            activity.getBinding().backToLiveButton.setVisibility(View.VISIBLE);
            // Ensure the button is drawn on top of the ImageView
            activity.getBinding().backToLiveButton.bringToFront();
        });
    }

}

File: AISuite_QuickStart/app/src/main/java/com/zebra/aisuite_quickstart/kotlin/handlers/UIHandler.kt
// Copyright 2025 Zebra Technologies Corporation and/or its affiliates. All rights reserved.
package com.zebra.aisuite_quickstart.kotlin.handlers

import android.annotation.SuppressLint
import android.content.SharedPreferences
import android.content.pm.ActivityInfo
import android.graphics.Bitmap
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.annotation.NonNull
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCaptureException
import androidx.camera.core.ImageProxy
import androidx.core.content.edit
import androidx.core.view.isVisible
import com.zebra.aisuite_quickstart.filtertracker.FilterDialog
import com.zebra.aisuite_quickstart.kotlin.CameraXLivePreviewActivity
import com.zebra.aisuite_quickstart.kotlin.camera.CameraManager
import com.zebra.aisuite_quickstart.utils.CommonUtils
import com.zebra.aisuite_quickstart.utils.CommonUtils.WAREHOUSE_LOCALIZER
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * UIHandler manages UI-related operations including spinner setup,
 * overlay management, and orientation changes.
 */
class UIHandler(
    private val activity: CameraXLivePreviewActivity,
    private val cameraManager: CameraManager,
    private val sharedPreferences: SharedPreferences?,
) {
    companion object {
        private const val TAG = "UIHandler"

        // Model constants
        private const val BARCODE_DETECTION = "Barcode"
        private const val LEGACY_BARCODE_DETECTION = "Legacy Barcode"
        private const val TEXT_OCR_DETECTION = "OCR"
        private const val LEGACY_OCR_DETECTION = "Legacy OCR"
        private const val ENTITY_ANALYZER = "Tracker"
        private const val PRODUCT_RECOGNITION = "Product Recognition"
        private const val LEGACY_PRODUCT_RECOGNITION = "Legacy Product Recognition"
        private const val ENTITY_VIEW_FINDER = "Entity Viewfinder"
    }

    var selectedModel: String = BARCODE_DETECTION
    var isEntityViewFinder: Boolean = false
    var isSpinnerInitialized = false

    private var imageCapture: ImageCapture? = null
    var isInCaptureMode: Boolean = false
    private var currentCapture: Bitmap? = null
    private val executors: ExecutorService = Executors.newFixedThreadPool(2)

    fun setupSpinner() {
        val dataAdapter = getStringArrayAdapter()
        activity.binding.spinner.adapter = dataAdapter
        activity.binding.spinner.onItemSelectedListener = createSpinnerListener()
        setUpTrackerFilter()
    }

    fun setUpTrackerFilter() {
        activity.binding.trackerFilter.setOnClickListener {
            val dialog = FilterDialog(activity)
            dialog.setCallback { filters ->
                sharedPreferences?.edit {

                    for (option in filters) {
                        putBoolean(option.title, option.isChecked)
                    }
                }
                activity.isModelLoaded = false
                activity.clearGraphicOverlay()
                activity.stopAnalyzing()
                cameraManager.unbindAll()
                activity.disposeModels()
                cameraManager.bindPreviewAndAnalysis(activity.getPreviewSurfaceProvider())
                activity.bindAnalysisUseCase()

            }
            dialog.show()
        }
    }

    @NonNull
    private fun getStringArrayAdapter(): ArrayAdapter<String> {
        val options = listOf(
            BARCODE_DETECTION,
            TEXT_OCR_DETECTION,
            PRODUCT_RECOGNITION,
            ENTITY_ANALYZER,
            ENTITY_VIEW_FINDER,
            WAREHOUSE_LOCALIZER,
            // LEGACY_BARCODE_DETECTION, // Uncomment to use barcode legacy option
            // LEGACY_OCR_DETECTION // Uncomment to use OCR legacy option
            //LEGACY_PRODUCT_RECOGNITION // Uncomment to use Product recognition legacy option
        )

        return ArrayAdapter(
            activity,
            com.zebra.aisuite_quickstart.R.layout.spinner_style,
            options
        ).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
    }

    private fun createSpinnerListener(): AdapterView.OnItemSelectedListener {
        return object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View?, pos: Int, id: Long) {
                selectedModel = adapterView.getItemAtPosition(pos).toString()
                isEntityViewFinder = selectedModel == ENTITY_VIEW_FINDER

                Log.e(TAG, "Selected option is $selectedModel")
                activity.binding.trackerFilter.isVisible = TextUtils.equals(selectedModel, ENTITY_ANALYZER)
                // Lock orientation when Entity Viewfinder is selected
                if (isEntityViewFinder) {
                    activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LOCKED
                    Log.d(TAG, "Orientation locked for Entity Viewfinder mode")
                } else {
                    activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
                    Log.d(TAG, "Orientation unlocked for $selectedModel mode")
                }
                activity.isModelLoaded = false

                // Show capture layout for supported models
                if (selectedModel.equals(BARCODE_DETECTION, ignoreCase = true) ||
                    selectedModel.equals(TEXT_OCR_DETECTION, ignoreCase = true) ||
                    selectedModel.equals(PRODUCT_RECOGNITION, ignoreCase = true) ||
                    selectedModel.equals(ENTITY_ANALYZER, ignoreCase = true) ||
                    selectedModel.equals(WAREHOUSE_LOCALIZER, ignoreCase = true)
                ) {
                    initializeCaptureFeature()
                    activity.binding.captureLayout.visibility = View.VISIBLE
                } else {
                    activity.binding.captureLayout.visibility = View.GONE
                }
                // Clear overlays and rebind the camera
                if (!isSpinnerInitialized) {
                    isSpinnerInitialized = true
                } else {
                    activity.clearGraphicOverlay()
                    activity.stopAnalyzing()
                    cameraManager.unbindAll()
                    activity.disposeModels()
                    cameraManager.bindPreviewAndAnalysis(activity.getPreviewSurfaceProvider())
                    activity.bindAnalysisUseCase()
                }
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {
                // No action needed
            }
        }
    }
    private fun initializeCaptureFeature() {
        activity.binding.captureButton.setOnClickListener {
            if (selectedModel.equals(BARCODE_DETECTION, ignoreCase = true) ||
                selectedModel.equals(TEXT_OCR_DETECTION, ignoreCase = true) ||
                selectedModel.equals(PRODUCT_RECOGNITION, ignoreCase = true) ||
                selectedModel.equals(ENTITY_ANALYZER, ignoreCase = true) ||
                selectedModel.equals(WAREHOUSE_LOCALIZER, ignoreCase = true)
            ) {
                if (activity.isModelLoaded) {
                    activity.binding.captureLayout.visibility = View.GONE
                    imageCapture = cameraManager.imageCapture
                    activity.stopAnalyzing()
                    activity.binding.control.visibility = View.GONE

                    executors.submit {
                        when {
                            selectedModel.equals(BARCODE_DETECTION, ignoreCase = true) ->
                                initializeCaptureDecoderAndCapture()
                            selectedModel.equals(TEXT_OCR_DETECTION, ignoreCase = true) ->
                                initializeCaptureOCRAndCapture()
                            selectedModel.equals(PRODUCT_RECOGNITION, ignoreCase = true) ->
                                initializeCaptureRecognitionAndCapture()
                            selectedModel.equals(ENTITY_ANALYZER, ignoreCase = true) ->
                                initializeCaptureTracker()
                            selectedModel.equals(WAREHOUSE_LOCALIZER, ignoreCase = true) ->
                                initializeCaptureWareHouseLocalizer()
                        }
                    }
                } else {
                    Toast.makeText(activity, "Model is not loaded", Toast.LENGTH_SHORT).show()
                }
            }
        }

        activity.binding.backToLiveButton.setOnClickListener {
            activity.binding.control.visibility = View.VISIBLE
            returnToLivePreview()
        }
    }

    private fun initializeCaptureDecoderAndCapture() {
        activity.initializeCaptureDecoder(::performCapture)
    }

    private fun initializeCaptureOCRAndCapture() {
        activity.initializeCaptureOCR(::performCapture)
    }

    private fun initializeCaptureRecognitionAndCapture() {
        activity.initializeCaptureRecognition(::performCapture)
    }

    private fun initializeCaptureTracker() {
        activity.initializeCaptureTracker(::performCapture)
    }

    private fun initializeCaptureWareHouseLocalizer() {
        activity.initializeCaptureWareHouse(::performCapture)
    }

    private fun performCapture() {
        if (imageCapture == null || isInCaptureMode) return
        Log.d(TAG, "Performing capture")

        imageCapture!!.takePicture(executors, object : ImageCapture.OnImageCapturedCallback() {
            override fun onCaptureSuccess(imageProxy: ImageProxy) {
                try {
                    Log.e(TAG, "Image Dims w x h = ${imageProxy.width} x ${imageProxy.height}")
                    Log.e(TAG, "Selected model: $selectedModel")
                    currentCapture = CommonUtils.rotateBitmapIfNeeded(imageProxy)
                    activity.runOnUiThread {
                        switchToCaptureMode()
                        activity.binding.capturedImageView.setImageBitmap(currentCapture)
                    }
                    when {
                        selectedModel.equals(TEXT_OCR_DETECTION, ignoreCase = true) ->
                            activity.processCaptureOCR(imageProxy)
                        selectedModel.equals(BARCODE_DETECTION, ignoreCase = true) ->
                            activity.processBarcodeWithCaptureDecoder(imageProxy)
                        selectedModel.equals(PRODUCT_RECOGNITION, ignoreCase = true) ->
                            activity.processCaptureRecognition(imageProxy)
                        selectedModel.equals(ENTITY_ANALYZER, ignoreCase = true) ->
                            activity.processCaptureTracker(imageProxy)
                        selectedModel.equals(WAREHOUSE_LOCALIZER, ignoreCase = true) ->
                            activity.processCaptureWareHouseLocalizer(imageProxy)
                    }
                    Log.d(TAG, "bitmap captured")
                } catch (e: Exception) {
                    Log.e(TAG, "Exception occurred while capturing: ${e.message}")
                }
            }

            override fun onError(exception: ImageCaptureException) {
                Log.e(TAG, "Capture operation failed", exception)
            }
        })
    }

    @SuppressLint("ResourceAsColor")
    private fun returnToLivePreview() {
        activity.runOnUiThread {
            isInCaptureMode = false
            currentCapture = null

            activity.clearGraphicOverlay()

            activity.binding.capturedImageView.visibility = View.GONE
            activity.binding.capturedImageView.setImageBitmap(null)
            activity.binding.backToLiveButton.visibility = View.GONE
            activity.binding.trackerFilter.isVisible = TextUtils.equals(selectedModel, ENTITY_ANALYZER)
            activity.binding.previewView.visibility = View.VISIBLE
            activity.binding.previewView.setBackgroundColor(android.R.color.black)
            activity.binding.graphicOverlay.visibility = View.VISIBLE
            activity.binding.captureLayout.visibility = View.VISIBLE
        }

        cameraManager.unbindAll()
        cameraManager.bindPreviewAndAnalysis(activity.getPreviewSurfaceProvider())

        when {
            selectedModel.equals(TEXT_OCR_DETECTION, ignoreCase = true) ->
                activity.setOCRAnalysis()

            selectedModel.equals(BARCODE_DETECTION, ignoreCase = true) ->
                activity.setBarcodeAnalysis()

            selectedModel.equals(PRODUCT_RECOGNITION, ignoreCase = true) -> {
                    activity.clearCaptureListener()
                    activity.setRecognitionAnalysis()
            }
            selectedModel.equals(ENTITY_ANALYZER, ignoreCase = true) ->
                activity.setTrackerAnalysis()
            selectedModel.equals(WAREHOUSE_LOCALIZER, ignoreCase = true) ->
                activity.setWareHouseAnalysis()
        }
    }

    private fun switchToCaptureMode() {
        activity.runOnUiThread {
            isInCaptureMode = true

            cameraManager.unbindImageAnalysis()

            activity.binding.previewView.visibility = View.GONE
            activity.binding.graphicOverlay.visibility = View.GONE
            activity.binding.captureLayout.visibility = View.GONE

            activity.binding.capturedImageView.visibility = View.VISIBLE
            activity.binding.backToLiveButton.visibility = View.VISIBLE
            activity.binding.backToLiveButton.bringToFront()
        }
    }


    fun updatePreviewVisibility(isEntityViewFinder: Boolean) {
        activity.binding.previewView.visibility = if (isEntityViewFinder) View.GONE else View.VISIBLE
        activity.binding.entityView.visibility = if (isEntityViewFinder) View.VISIBLE else View.GONE
    }

    /**
     * Enable or disable the spinner and change its appearance
     * @param enabled true to enable spinner, false to disable
     */
    fun setSpinnerEnabled(enabled: Boolean) {
        activity.runOnUiThread {
            activity.binding.spinner.let { spinner ->
                spinner.isEnabled = enabled
                spinner.alpha = if (enabled) 1.0f else 0.5f

            }
            activity.binding.trackerFilter.isEnabled = enabled
            activity.binding.trackerFilter.alpha = if (enabled) 1.0f else 0.5f

            activity.binding.captureButton.isEnabled = enabled
            activity.binding.captureButton.alpha = if (enabled) 1.0f else 0.5f
        }
    }
}